### ISSUE TITLE
Matching Lists with Iterables

### ISSUE THREAD

**Issue by tomaine2002** [dart-lang/mockito#186]
The [DeepCollectionEquality](https://pub.dartlang.org/documentation/collection/latest/collection/DeepCollectionEquality-class.html) class used for argument matching intentionally does not consider a List equivalent to a non-list Iterable, even if they have the same elements in the same order.

Generally, the test isn't interested in whether the system under test passes a List or some other Iterable. Instead, the test just wants to verify the Iterable has the expected elements in the expected order.

The simple way to specify the expected iterable argument is to use a List with the expected values, in the expected order. However, if the system under test passes an Iterable that is not a list, the stub will never match.

**Comment by natebosch** [comment]
Can you give more specifics on the usage you'd like to enable here? Is it something like `when(mock.someCall([1,2,3]))` ?

Can you try `when(mock.someCall(argThat(orderedEquals([1,2,3]))))`?

**Comment by tomaine2002** [comment]
>  you give more specifics on the usage you'd like to enable here? Is it something like when(mock.someCall([1,2,3])) ?

Yes, that's the way I'd like to set up the stub. It seems to me this should match for any Iterable with the elements 1, 2, 3 in that order.

> Can you try when(mock.someCall(argThat(orderedEquals([1,2,3]))))?

That works as expected. Thanks for this suggestion. It seems unintuitive to me that `when(mock.someCall([1,2,3]))` and `when(mock.someCall(argThat(orderedEquals([1,2,3]))))` do not have the same meaning.

I'd prefer the simpler `when(mock.someCall([1,2,3]))` if it had the same effect as `when(mock.someCall(argThat(orderedEquals([1,2,3]))))`

The current behavior will likely lead to tests that are too tight. For example, if the system under test  passes a List for an Iterable argument and the stub matches with a List like `when(mock.someCall([1,2,3]))` everything initially works. However, if someone later changes the system under test to use a non-list Iterable (e.g. replacing `someCall(list)` with `someCall(list.where(...))`), then the test will start to fail because the stub doesn't match, even if the system under test is correct.
